File: plugins/collector/event/apps/web/jetty.go
package web

import (
	"collector/cache/process"
	"collector/event/apps"
	"path/filepath"
	"regexp"
	"strings"
)

type Jetty struct {
	version string
	reg     *regexp.Regexp
}

func (Jetty) Name() string { return "jetty" }

func (Jetty) Type() string { return "web" }

func (j Jetty) Version() string { return j.version }

func (j *Jetty) Match(p *process.Process) bool {
	if p.Name != "java" {
		return false
	}
	fds, err := p.Fds()
	if err != nil {
		return false
	}
	if j.reg == nil {
		j.reg = regexp.MustCompile(`jetty-client-(\d+\.)+(\d+)\.jar`)
	}
	for _, fd := range fds {
		if filepath.Ext(fd) != ".jar" {
			continue
		}
		if s := j.reg.FindString(filepath.Base(fd)); s != "" {
			j.version = strings.TrimSuffix(strings.TrimPrefix(s, "jetty-client-"), ".jar")
			return true
		}
	}
	return false
}

func (n *Jetty) Run(p *process.Process) (m map[string]string, err error) {
	return
}

func init() { apps.Regist(&Jetty{}) }

File: plugins/collector/src/event/apps/web/jetty.rs
use crate::event::apps::{find_version, jar_names_for_pid, AppProc, IApp};
use std::collections::HashMap;

pub struct Jetty {
    version: String,
}

impl Jetty {
    pub fn new() -> Self { Jetty { version: String::new() } }
}

impl IApp for Jetty {
    fn name(&self)     -> &'static str { "jetty" }
    fn app_type(&self) -> &'static str { "web" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool {
        p.name == "java" && p.argv.contains("jetty")
    }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        self.version.clear();
        // "jetty-client-9.4.48.v20220622.jar"
        for jar in jar_names_for_pid(p.pid) {
            if jar.starts_with("jetty-client-") && jar.ends_with(".jar") {
                let mid = jar.trim_start_matches("jetty-client-").trim_end_matches(".jar");
                let ver = find_version(mid);
                if !ver.is_empty() { self.version = ver; break; }
            }
        }
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
